### PR TITLE
MMF-999 Add management of quality gates on organizations

### PR DESCRIPTION
**Pull request by ehartmann**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/organization/OrganizationCreationImpl.java [hunk 1]
```diff
@@ -180,15 +179,16 @@ private void validate(NewOrganization newOrganization) {
...
 
-  private OrganizationDto insertOrganization(DbSession dbSession, NewOrganization newOrganization, Consumer<OrganizationDto> extendCreation) {
+  private OrganizationDto insertOrganization(DbSession dbSession, NewOrganization newOrganization, Consumer<OrganizationDto>... extendCreation) {
     OrganizationDto res = new OrganizationDto()
       .setUuid(uuidFactory.create())
       .setName(newOrganization.getName())
       .setKey(newOrganization.getKey())
       .setDescription(newOrganization.getDescription())
       .setUrl(newOrganization.getUrl())
+      // TODO .setDefaultQualityGateUuid("" + qualityGateFinder.getBuiltInQualityGate(dbSession).getId())
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/organization/OrganizationCreationImpl.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO][INFO] [QualityGates.java#L97](https://github.com/SonarSource/sonarqube/blob/04f036814694ce0dc44421b37d97016c54182951/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java#L97): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)
1. ![INFO][INFO] [QualityGates.java#L112](https://github.com/SonarSource/sonarqube/blob/04f036814694ce0dc44421b37d97016c54182951/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGates.java#L112): Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWsSupport.java [hunk 2]
```diff
@@ -49,10 +56,32 @@ QualityGateConditionDto getCondition(DbSession dbSession, long id) {
...
+      .setRequired(false)
+      .setInternal(false)
+      .setExampleValue("my-org");
+  }
+
+  /**
+   * @deprecated use {@link #getActions(OrganizationDto, QualityGateDto, QualityGateDto)} instead
+   */
+  @Deprecated
   Qualitygates.Actions getActions(QualityGateDto qualityGate, @Nullable QualityGateDto defaultQualityGate) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWsSupport.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWsSupport.java [hunk 3]
```diff
@@ -49,10 +56,32 @@ QualityGateConditionDto getCondition(DbSession dbSession, long id) {
     return checkFound(dbClient.gateConditionDao().selectById(id, dbSession), "No quality gate condition with id '%d'", id);
   }
 
+  /**
+   * @deprecated use {@link #isQualityGateAdmin(OrganizationDto)} instead
+   */
+  @Deprecated
   boolean isQualityGateAdmin() {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWsSupport.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/ShowAction.java [hunk 4]
```diff
@@ -87,12 +91,23 @@ public void handle(Request request, Response response) {
...
+      writeProtobuf(buildResponse(organization, qualityGate, defaultQualityGate.orElse(null), conditions, metricsById), request, response);
+    }
+  }
+
+  private QualityGateDto getByNameOrId(DbSession dbSession, OrganizationDto organization, @Nullable String name, @Nullable Long id) {
+    if (name != null) {
+      return checkFound(dbClient.qualityGateDao().selectByOrganizationAndName(dbSession, organization, name), "No quality gate has been found for name %s", name);
+    }
+    if (id != null) {
+      return qualityGateFinder.getById(dbSession, id);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/ShowAction.java]: MINOR Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java [hunk 5]
```diff
@@ -57,18 +59,16 @@ public QualityGateFinder(DbClient dbClient) {
     }
   }
 
+  /**
+   * @deprecated Use {@link #getByOrganizationAndId(DbSession, OrganizationDto, long)}
+   */
+  @Deprecated
   public QualityGateDto getById(DbSession dbSession, long qualityGateId) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWsSupport.java [hunk 6]
```diff
@@ -68,12 +86,29 @@ boolean isQualityGateAdmin() {
...
+      .orElseGet(() -> defaultOrganizationProvider.get().getKey());
+    Optional<OrganizationDto> organizationDto = dbClient.organizationDao().selectByKey(dbSession, organizationKey);
+    return checkFoundWithOptional(organizationDto, "No organization with key '%s'", organizationKey);
+  }
+
+  /**
+   * @deprecated use {@link #checkCanEdit(QGateWithOrgDto)} instead
+   */
+  @Deprecated
   void checkCanEdit(QualityGateDto qualityGate) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/QualityGatesWsSupport.java]: INFO Do not forget to remove this deprecated code someday. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1133)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QualityGateFinder.java#L52](https://github.com/SonarSource/sonarqube/blob/a78dce23cac00887350b8e3dcc43f4d011dbebf1/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L52): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QualityGateFinder.java#L52](https://github.com/SonarSource/sonarqube/blob/79ffd2b05799f027994b7dd678d7ab6b865f08ca/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L52): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QualityGateFinder.java#L52](https://github.com/SonarSource/sonarqube/blob/bb125a760f532bb3703d491c3b7d7cb3139121ad/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L52): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QualityGateFinder.java#L52](https://github.com/SonarSource/sonarqube/blob/5a86655bd71f4f345dbe04462ac4d95afabeb12b/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L52): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QualityGateFinder.java#L52](https://github.com/SonarSource/sonarqube/blob/b3b6d0552640ead23263e6556f7b20a2d10aad96/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L52): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QualityGateFinder.java#L52](https://github.com/SonarSource/sonarqube/blob/4a194e1799a2ac81109f25682c2b582ed7536076/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L52): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MINOR][MINOR] 2 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [QualityGateFinder.java#L52](https://github.com/SonarSource/sonarqube/blob/bb95879f135ea8fe9096ed0233a2e92f9b4035b2/server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java#L52): Remove this use of "getById"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-db-dao/src/test/java/org/sonar/db/organization/OrganizationDaoTest.java [hunk 7]
```diff
@@ -1020,7 +1020,7 @@ private OrganizationDto copyOf(OrganizationDto organizationDto) {
       .setName(organizationDto.getName())
       .setDescription(organizationDto.getDescription())
       .setUrl(organizationDto.getUrl())
-      // TODO : .setDefaultQualityGateUuid(dbClient.qualityGateDao().selectBuiltIn(dbSession).getId().toString())
+      .setDefaultQualityGateUuid(organizationDto.getDefaultQualityGateUuid())
```
Comment (julienlancelot) [inline server/sonar-db-dao/src/test/java/org/sonar/db/organization/OrganizationDaoTest.java]: As the commit is only about existing quality gate, nothing should has been done on new one

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java [hunk 8]
```diff
@@ -57,20 +60,12 @@ public QualityGateFinder(DbClient dbClient) {
...
-    if (id != null) {
-      return getById(dbSession, id);
-    }
-    throw new IllegalArgumentException("No parameter has been set to identify a quality gate");
+  public QGateWithOrgDto getByOrganizationAndId(DbSession dbSession, OrganizationDto organization, long qualityGateId) {
+    return checkFound(dbClient.qualityGateDao().selectByOrganizationAndId(dbSession, organization, qualityGateId),
+      "No quality gate has been found for id %s in organization %s", qualityGateId, organization.getName());
   }
 
+  // TODO As there is always a default quality gate, this method should not return an optional
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/QualityGateFinder.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'

File: server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SearchAction.java [hunk 9]
```diff
@@ -43,47 +68,106 @@ public void define(WebService.NewController controller) {
...
+
+  private List<ProjectQgateAssociationDto> keepAuthorizedProjects(DbSession dbSession, List<ProjectQgateAssociationDto> projects) {
+    if (userSession.isRoot()) {
+      // the method AuthorizationDao#keepAuthorizedProjectIds() should be replaced by
+      // a call to UserSession, which would transparently support roots.
+      // Meanwhile root is explicitly handled.
+      return projects;
+    }
+    List<Long> projectIds = projects.stream().map(ProjectQgateAssociationDto::getId).collect(MoreCollectors.toList());
+    Collection<Long> authorizedProjectIds = dbClient.authorizationDao().keepAuthorizedProjectIds(dbSession, projectIds, userSession.getUserId(), UserRole.USER);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/qualitygate/ws/SearchAction.java]: MINOR Remove this use of "keepAuthorizedProjectIds"; it is deprecated. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**